### DISCUSSION TITLE
Avoiding "return to terminal" after cursive.run()

### DISCUSSION THREAD

**Discussion by massimo79m** [gyscos/cursive#755] [Q&A]
I'm developing a software with cursive, and it is made of various "forms".
It prepares a form, then it run() it, then it crunches the data and runs another form.
I tried all the backends, but when the run() ends, the form disappears and the previous terminal reappears(the terminal i had before launching my application)  between the end of the first run() and the next run().
It is quite annoying, for me would be optimal if the first form would remain on the screen while the second form is calculated.
Is there a way?

Thank you.

**Comment by gyscos** [comment]
Hi, and thanks for the question!

The easiest way would be to have a single call to run(). and replace the views there.

* First, build the first form.
* When the user "submits", maybe show a loading screen, but do not call `Cursive::quit()` yet. Start preparation for the new form in a new thread. Give this thread a clone of [cursive's callback sink](https://docs.rs/cursive/latest/cursive/struct.Cursive.html#method.cb_sink).
* When the second form is ready, use the callback sink to send a callback to cursive: remove all the existing layers (or alternatively, [start a new "screen"](https://docs.rs/cursive/latest/cursive/struct.Cursive.html#method.add_active_screen)), then add the new layers for this second form.
* Continue like this, only calling `Cursive::quit()` when you actually want to leave the application.

As a bonus, you could improve the loading screen: either show a spinner, or (if you have it) display the current progress (for example using [a progress bar](https://docs.rs/cursive/latest/cursive/views/struct.ProgressBar.html)), maybe a "cancel" button, ...
